identity/issuer/issuer.go

```diff
@@ -253,62 +365,74 @@ func (is *Issuer) ID() *core.ID {
 // SyncIdenStatePublic updates the IdenStateOnChain and IdenStatePending from
 // the values in the Smart Contract.
 func (is *Issuer) SyncIdenStatePublic() error {
-	idenStateData, err := is.idenStateReader.GetState(is.id)
+	is.rw.Lock()
+	defer is.rw.Unlock()
+	tx, err := is.storage.NewTx()
 	if err != nil {
 		return err
 	}
-	// If there's an idenStatePending, verify that the result from the
-	// smart contract matches.  Otherwise, the result must be the
-	// idenStateOnChain.
-	if is.idenStatePending != nil {
-		if !idenStateData.IdenState.Equals(is.idenStatePending) {
-			return fmt.Errorf("Fatal error: Identity State in the Smart Contract (%v)"+
-				" doesn't match the expected Pending one (%v).",
-				idenStateData.IdenState, is.idenStatePending)
-		}
-	} else {
-		if !idenStateData.IdenState.Equals(is.idenStateOnChain) {
-			return fmt.Errorf("Fatal error: Identity State in the Smart Contract (%v)"+
-				" doesn't match the expected OnChain one (%v).",
-				idenStateData.IdenState, is.idenStatePending)
+	if err := is.syncIdenStatePublic(tx); err != nil {
+		return err
+	}
+	if err := tx.Commit(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (is *Issuer) syncIdenStatePublic(tx db.Tx) error {
+	if is.idenPubOnChain == nil {
+		return ErrIdenPubOnChainNil
+	}
+	idenStateData, err := is.idenPubOnChain.GetState(is.id)
+	if err != nil {
+		return err
+	}
+	if is.idenStatePending().Equals(&merkletree.HashZero) {
+		// If there's no IdenState pending to be set on chain, the
+		// obtained one must be the idenStateOnChain (Zero for genesis
+		// / empty in the smart contract).
+		if idenStateData.IdenState.Equals(is.idenStateOnChain()) {
+			return nil
 		}
+
+		return fmt.Errorf("Fatal error: Identity State in the Smart Contract (%v)"+
+			" doesn't match the expected OnChain one (%v).",
+			idenStateData.IdenState, is.idenStateOnChain())
 	}
+	// If there's an IdenState pending to be set on chain, the
+	// obtained one can be:
 
-	is.idenStatePending = nil
-	is.idenStateOnChain = idenStateData.IdenState
+	// a. the idenStatePending (in this case, we no longer have an
+	// IdenState pending and it becomes the idenStateOnChain).
+	if idenStateData.IdenState.Equals(is.idenStatePending()) {
```

Comment:
Suggestion:
* since the tx is only used in this case, declare it and use it here.
* split logic: ceck if synced + update (if not synced)